File: Data Structures & Algorithms/count-number-of-islands/submission-12.go
func numIslands(grid [][]byte) int {

    ROWS, COLS := len(grid), len(grid[0])
    
    islands := 0

    for r := 0; r < ROWS; r++ {
        for c:= 0; c < COLS; c++ {
            if grid[r][c] == '1'{
                islands += bfs(r, c, grid)
            }
        }
    }
    return islands
}

func bfs(r, c int, grid [][]byte) int {
    ROWS, COLS := len(grid), len(grid[0])
    island := 0
    q := [][]int{{r, c}}
    dirs := [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
    for len(q)> 0{
        r, c := q[0][0], q[0][1]
        grid[r][c] ='0'
        q = q[1:]
        for _, coord := range dirs {
            dr, dc := coord[0], coord[1]
            nr, nc := r + dr, c + dc
            if nr < 0 || nc < 0 || nr == ROWS || nc == COLS || grid[nr][nc] == '0' {
                continue
            }

            q = append(q, []int{nr, nc})
            grid[nr][nc] = '0'

        }
    }
    island++
    return island
}

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-12.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        if grid[0][0] == 1 or grid[ROWS-1][COLS-1] == 1:
            return -1
        
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, 1), (1, -1), (-1, -1)]
        q = deque()
        visited = set()
        q.append((0, 0))

        length = 0
        while q:
            for _ in range(len(q)):
                r, c = q.popleft()
                if min(r, c) < 0 or r == ROWS or c == COLS or grid[r][c] == 1 or (r, c) in visited:
                    continue
                if r == ROWS -1 and c == COLS - 1:
                    return length + 1
                visited.add((r, c))
                for dr, dc in dirs:
                    q.append((r + dr, c + dc))
            length += 1
        return -1

            
